sgx-hasher/src/lib.rs

```diff
@@ -77,6 +81,68 @@ impl Hasher {
     }
 }
 
+pub fn to_array(bn: bn::BigNum) -> [u8; 384] {
+    let mut array = [0u8; 384];
+    array.copy_from_slice(&bn.to_vec());
+    array
+}
+
+pub fn sign(hash: &[u8; 32], key: &rsa::Rsa<Private>) -> Result<sig::Signature> {
+    let author = sig::Author::new(sig::Vendor::default(), 0u32, 0u32);
+
+    let contents = sig::Contents::new(
+        misc::MiscSelect::EXINFO,
+        attr::Attributes::new(attr::Flags::MODE_64_BIT, attr::Xfrm::all()),
+        *hash,
+        0,
+        0,
+    );
+
+    let rsa_key = PKey::from_rsa(key.to_owned())?;
+    let md = hash::MessageDigest::sha256();
+    let mut signer = sign::Signer::new(md, &rsa_key)?;
+    signer.update(author.as_ref())?;
+    signer.update(contents.as_ref())?;
+    let signature = signer.sign_to_vec()?;
+
+    let e = hex::encode(key.e().to_vec());
+    let e = u32::from_str_radix(&e[..], 16).unwrap();
+
+    let m = key.n();
+
+    let s = bn::BigNum::from_slice(&signature)?;
+
+    let mut ctx = bn::BigNumContext::new()?;
+
+    let mut q1 = bn::BigNum::new()?;
+    let mut q2 = bn::BigNum::new()?;
+    let mut qr = bn::BigNum::new()?;
+    let mut s_2 = bn::BigNum::new()?;
+    let mut s_3 = bn::BigNum::new()?;
+    let mut s_3minus = bn::BigNum::new()?;
+    let mut sigmod = bn::BigNum::new()?;
+    let mut sigmodq1 = bn::BigNum::new()?;
+
+    s_2.sqr(&s, &mut ctx)?; // sig^2
+    q1.div_rem(&mut qr, &s_2, &m, &mut ctx)?; // floor(sig^2 / modulus)
+
+    s_3.checked_mul(&s, &s_2, &mut ctx)?; // sig^3
+    sigmod.checked_mul(&s, &m, &mut ctx)?; // sig * modulus
+    sigmodq1.checked_mul(&q1, &sigmod, &mut ctx)?; // sig * modulus * q1
+    s_3minus.checked_sub(&s_3, &sigmodq1)?; // sig^3 - (q1 * sig * modulus)
+    q2.div_rem(&mut qr, &s_3minus, &m, &mut ctx)?; // floor((sig^3 - q1 * sig * modulus) / modulus)
```

Comment:
I don't think this algorithm is correct. See: https://github.com/jsakkine-intel/linux-sgx/blob/master/tools/testing/selftests/x86/sgx/sgxsign.c#L362

Also, I'd really prefer to use the Rust ops. The `openssl` crate appears to implement them. But I couldn't get them to work. That's probably something to pursue with upstream.

Reply:
I'm trying to implement this the way it is implemented here https://github.com/jsakkine-intel/linux-sgx/blob/master/tools/testing/selftests/x86/sgx/sgxsign.c#L362. But their algorithm seems to say `s * qr == s^3 - q1 * s * m`, and I don't understand why this must be true. 

By Rust ops do you mean the regular `*` and `/` operators? Should I extract `u32` from each `bignum` to do the calculations? Or did you want me to do an upstream fix with `openssl` as a prereq for this PR?

Reply:
I will just adapt it from what you had, although I'm still not sure about the algorithm and I would like to understand it before pushing it.